mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala

```diff
@@ -0,0 +1,141 @@
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  override def transformSchema(schema: StructType): StructType = {
+    SchemaUtils.checkColumnType(schema, $(inputCol), DoubleType)
+    val inputFields = schema.fields
+    require(inputFields.forall(_.name != $(outputCol)),
+      s"Output column ${$(outputCol)} already exists.")
+    val attr = NominalAttribute.defaultAttr.withName($(outputCol))
+    val outputFields = inputFields :+ attr.toStructField()
+    StructType(outputFields)
+  }
+
+  override def fit(dataset: DataFrame): Bucketizer = {
+    val input = dataset.select($(inputCol)).map { case Row(feature: Double) => feature }
+    val samples = getSampledInput(input, $(numBuckets))
+    val splits = Array(Double.NegativeInfinity) ++ findSplits(samples, $(numBuckets) - 1) ++
+      Array(Double.PositiveInfinity)
+    val bucketizer = new Bucketizer(uid).setSplits(splits)
+    copyValues(bucketizer)
+  }
+
+  /**
+   * Sampling from the given dataset to collect quantile statistics.
+   */
+  private def getSampledInput(dataset: RDD[Double], numBins: Int): Array[Double] = {
+    val totalSamples = dataset.count()
+    assert(totalSamples > 0)
+    val requiredSamples = math.max(numBins * numBins, 10000)
+    val fraction = math.min(requiredSamples / dataset.count(), 1.0)
+    dataset.sample(withReplacement = false, fraction, new XORShiftRandom().nextInt()).collect()
+  }
+
+  /**
+   * Compute split points with respect to the sample distribution.
+   */
+  private def findSplits(samples: Array[Double], numSplits: Int): Array[Double] = {
+    val valueCountMap = samples.foldLeft(Map.empty[Double, Int]) { (m, x) =>
+      m + ((x, m.getOrElse(x, 0) + 1))
+    }
+    val valueCounts = valueCountMap.toSeq.sortBy(_._1).toArray
+    val possibleSplits = valueCounts.length
+    if (possibleSplits <= numSplits) {
+      valueCounts.map(_._1)
+    } else {
+      val stride: Double = samples.length.toDouble / (numSplits + 1)
+      val splitsBuilder = mutable.ArrayBuilder.make[Double]
+      var index = 1
+      // currentCount: sum of counts of values that have been visited
+      var currentCount = valueCounts(0)._2
+      // targetCount: target value for `currentCount`.
+      // If `currentCount` is closest value to `targetCount`,
+      // then current value is a split threshold.
+      // After finding a split threshold, `targetCount` is added by stride.
+      var targetCount = stride
+      while (index < valueCounts.length) {
+        val previousCount = currentCount
+        currentCount += valueCounts(index)._2
+        val previousGap = math.abs(previousCount - targetCount)
+        val currentGap = math.abs(currentCount - targetCount)
+        // If adding count of current value to currentCount
+        // makes the gap between currentCount and targetCount smaller,
+        // previous value is a split threshold.
+        if (previousGap < currentGap) {
+          splitsBuilder += valueCounts(index - 1)._1
```

Comment:
I'm worried that this will not work when the final valueCounts bucket has a lot of values.  In that case, in the final iteration where index = valueCounts.length - 1, previousGap will be > currentGap, and we will never add the last value (valueCounts.last._1).  A test with numBuckets = 3 and values
1,2,3,3,3,3,3,3,3 might catch this, though you might need to play around with the numbers.